src/overlay005/field_control.c

```diff
@@ -0,0 +1,1087 @@
+static BOOL Field_ProcessStep(FieldSystem *fieldSystem);
+static BOOL Field_CheckCoordEvent(FieldSystem *fieldSystem);
+static BOOL Field_CheckTransition(FieldSystem *fieldSystem, const int playerX, const int playerZ, const u8 curTileBehavior);
+static BOOL Field_UpdateDaycare(FieldSystem *fieldSystem);
+static BOOL Field_UpdatePoison(FieldSystem *fieldSystem);
+static BOOL Field_UpdateSafari(FieldSystem *fieldSystem);
+static BOOL Field_UpdateVsSeeker(FieldSystem *fieldSystem);
+static BOOL Field_UpdatePokeRadar(FieldSystem *fieldSystem);
+static BOOL Field_CheckSign(FieldSystem *fieldSystem);
+static BOOL Field_UpdateRepel(FieldSystem *fieldSystem);
+static BOOL Field_UpdateFriendship(FieldSystem *fieldSystem);
+static void Field_CalculateFriendship(FieldSystem *fieldSystem);
+static void Field_PlayerPos(const FieldSystem *fieldSystem, int *playerX, int *playerZ);
+static void Field_Step(const FieldSystem *fieldSystem, int *playerX, int *playerZ);
+static void Field_StepInDirection(const FieldSystem *fieldSystem, int playerDir, int *playerX, int *playerZ);
+static u8 Field_CurrentTileBehavior(const FieldSystem *fieldSystem);
+static u8 Field_NextTileBehavior(const FieldSystem *fieldSystem);
+static BOOL Field_MapConnection(const FieldSystem *fieldSystem, int playerX, int playerZ, Location *nextMap);
+static void Field_TrySetMapConnection(FieldSystem *fieldSystem);
+static BOOL Field_DistortionInteract(FieldSystem *fieldSystem, MapObject **object);
+static int Field_CheckTrainerInfo(void);
+
+static void FieldInput_Clear (FieldInput *input)
+{
+    input->interact = FALSE;
+    input->endMovement = FALSE;
+    input->menu = FALSE;
+    input->registeredItem = FALSE;
+    input->sign = FALSE;
+    input->mapTransition = FALSE;
+    input->movement = FALSE;
+    input->dummy1 = FALSE;
+    input->dummy2 = FALSE;
+    input->dummy3 = FALSE;
+    input->dummy4 = FALSE;
+    input->dummy5 = FALSE;
+    input->playerDir = DIR_NONE;
+    input->transitionDir = DIR_NONE;
+}
+
+void FieldInput_Update (FieldInput *input, FieldSystem *fieldSystem, u16 pressedKeys, u16 heldKeys)
+{
+    FieldInput_Clear(input);
+
+    int moveState = Player_MoveState(fieldSystem->playerAvatar);
+    int avatarState = PlayerAvatar_MoveState(fieldSystem->playerAvatar);
+    int playerDir = PlayerAvatar_GetDir(fieldSystem->playerAvatar);
+
+    input->pressedKeys = pressedKeys;
+    input->heldKeys = heldKeys;
+
+    if (moveState == PLAYER_MOVE_STATE_END || moveState == PLAYER_MOVE_STATE_NONE) {
+        if (pressedKeys & PAD_BUTTON_X) {
+            input->menu = TRUE;
+        }
+
+        if (pressedKeys & PAD_BUTTON_Y) {
+            input->registeredItem = TRUE;
+        }
+
+        if (pressedKeys & PAD_BUTTON_A) {
+            input->interact = TRUE;
+        }
+
+        if (pressedKeys & PAD_BUTTON_B) {
+            input->dummy1 = TRUE;
+        }
+
+        if (heldKeys & (PAD_KEY_UP | PAD_KEY_DOWN | PAD_KEY_LEFT | PAD_KEY_RIGHT)) {
```

Comment:
Should these be joined into a generic `PAD_KEY` disjunction? Maybe in `core_sys.h`